### PR TITLE
Discord OAuth part 1: Schema change, extended prisma client, discord oauth stubs

### PR DESCRIPTION
**Pull request by cherylli**
# Description

### Part 1 of discord oauth: 
1. Schema change (fixed seed and tests which reference the old schema)
2. extended prisma client to make it reusable for future oauth providers, we can also add more "extensions" in future now this is setup 
3. Setup passport discord strategy and guard, and stubs, console.logs will be removed in the next part

### To test this PR
1. obtain discord oauth client id and client secret (on clickup, you can also make your own for dev), and put them in .env
2. `npx prisma db push` for both dev and test db
3. go to `http://localhost:8000/api/v1/auth/discord/login` a discord authorization table will pop up, and it should return the user's basic discord profile 

## Issue link

[Clickup Link](https://app.clickup.com/t/86azr0q2z)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Feature updates / changes
- [ ] Tests
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have updated the change log

### CODE REVIEW

File: src/auth/strategies/discord.strategy.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        @Inject("DISCORD_OAUTH")
+        private readonly discordAuthService: IAuthProvider,
+    ) {
+        super({
+            clientID: process.env.DISCORD_CLIENT_ID,
+            clientSecret: process.env.DISCORD_CLIENT_SECRET,
+            callbackURL: process.env.DISCORD_CALLBACK_URL,
+            scope: ["identify"],
+        });
+    }
```
Comment (timDeHof) [inline src/auth/strategies/discord.strategy.ts:19]: I'm getting an error saying `context.$queryRaw is not a function`
Screenshot 2024-08-11 at 6 48 27 PM

File: src/auth/strategies/discord.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { Inject, Injectable } from "@nestjs/common";
+import { PassportStrategy } from "@nestjs/passport";
+import { Strategy } from "passport-discord";
+import { Profile } from "passport-discord";
```
Comment (timDeHof) [inline src/auth/strategies/discord.strategy.ts:4]:
```suggestion
import { Strategy, Profile } from "passport-discord";
```
you can combine the imports. unless you wanted to import the `profile` types from "passport-discord"

File: src/prisma/prisma-extensions.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                providerUserId: string,
+            ) {
+                const context = Prisma.getExtensionContext(this);
+                return context.$queryRaw`
+                            SELECT userOAuthProfile.*
+                            FROM userOAuthProfile
+                            JOIN oAuthProviders ON userOAuthProfile.providerId = oAuthProviders.id
+                            WHERE oAuthProviders.name = ${providerName} AND userOAuthProfile.providerUserId = ${providerUserId}
+                        `;
+            },
```
Comment (timDeHof) [inline src/prisma/prisma-extensions.ts:18]: I got an error from the exceptionsHandler after login to discord with my own email and password. 
Screenshot 2024-08-11 at 6 48 27 PM

File: src/auth/strategies/discord.strategy.ts (hunk 2 above)
Comment (cherylli) [reply src/auth/strategies/discord.strategy.ts:4]: thanks, that's some weird auto import 😂

**Review by timDeHof — CHANGES_REQUESTED** [review]
I was able to go to the discord signin link but after logging in I got an error message about context.$queryRaw is not a function.

**Comment by cherylli** [comment]
thanks @timDeHof. It's fixed now. I should have extended query instead of model. 

Also fixed the actual raw query, if you add a row to UserOAuthProfile and OAuthProvider (name: 'discord') it should work

it will return 401 which is expected as I haven't implemented any authentication logic yet

**Comment by timDeHof** [comment]
> thanks @timDeHof. It's fixed now. I should have extended query instead of model.
> 
> Also fixed the actual raw query, if you add a row to UserOAuthProfile and OAuthProvider (name: 'discord') it should work
> 
> it will return 401 which is expected as I haven't implemented any authentication logic yet

yup, thats what I get.
image

**Comment by cherylli** [comment]
forgot to generate prisma migration - fixed now

**Review by JoshuaHinman — COMMENTED** [review]
image
I went through the steps and got this response - what would be a valid redirect URI to use?

**Comment by cherylli** [comment]
The redirect URL in clickup where we have all the .env should work. Unless you're using another port locally

**Comment by JoshuaHinman** [comment]
> The redirect URL in clickup where we have all the .env should work. Unless you're using another port locally

I'm actually running on 8001 locally- 8000 is already taken.  

I tried generating my own id, secret and redirect url on discord, which gets me an authorization window.

I tried redirecting to `/docs` ,which worked, but `/api/v1/auth/discord/login` returns 500

**Comment by cherylli** [comment]
@Get("/discord/redirect")

Which means api/v1/auth/discord/redirect

The thing is I only have port 8000 in the shared discord developer portal. It should work if you make your own discord oauth 

I could also try if I can add a second redirect URL when I get home tonight

**Comment by JoshuaHinman** [comment]
> ```
> @Get("/discord/redirect")
> ```
> 
> Which means api/v1/auth/discord/redirect

Ok - I used this redirect and got the expected 401

**Review by JoshuaHinman — APPROVED** [review]
Got it to work

**Comment by cherylli** [comment]
nice. I'm able to add another redirect uri in the discord oauth portal
image

**Review by timDeHof — APPROVED** [review]
looks good. I'm still received an 401 status code. but I did received the accessToken, RefreshToken, and my discord profile.

**Pull request merged**